sklearn/utils/_clibs.py

```diff
@@ -0,0 +1,411 @@
+        Maximum number of thread that can be used in thread pools
+
+        If int, sets the maximum number of thread to `limits` for each C-lib
+        selected by `subset`.
+
+        If dict(supported_libraries: max_threads), sets a custom maximum number
+        of thread for each C-lib.
+
+    subset : string or None, optional (default="all")
+        Subset of C-libs to limit. Used only if `limits` is an int
+
+        "all" : limit all supported C-libs.
+
+        "blas" : limit only BLAS supported C-libs.
+
+        "openmp" : limit only OpenMP supported C-libs. It can affect the number
+                   of threads used by the BLAS C-libs if they rely on OpenMP.
+
+    reload_clib : bool, (default=False)
+        If `reload_clib` is `True`, first loop through the loaded libraries to
+        ensure that this function is called on all available libraries.
+
+    Returns
+    -------
+    dynamic_threadpool_size : dict
+        contains pairs `('clib': boolean)` which are True if `clib` have been
+        found and can be used to scale the maximal number of threads
+        dynamically.
+    """
+    wrapper = _get_wrapper(reload_clib)
+    return wrapper.set_thread_limits(limits, subset)
+
+
+def get_thread_limits(reload_clib=True):
+    """Return maximal thread number for threadpools in supported C-lib
+
+    Parameters
+    ----------
+    reload_clib : bool, (default=True)
+        If `reload_clib` is `True`, first loop through the loaded libraries to
+        ensure that this function is called on all available libraries.
+
+    Returns
+    -------
+    thread_limits : dict
+        Contains the maximal number of threads that can be used in supported
+        libraries or None when the library is not available. The key of the
+        dictionary are "openmp_gnu", "openmp_intel", "openmp_win32",
+        "openmp_llvm", "openblas", "mkl" and "mkl_win32".
+    """
+    wrapper = _get_wrapper(reload_clib)
+    return wrapper.get_thread_limits()
+
+
+@contextmanager
+def thread_limits_context(limits=1, subset=None):
+    """Context manager for C-libs thread limits
+
+    Parameters
+    ----------
+    limits : int or dict, (default=1)
+        Maximum number of thread that can be used in thread pools
+
+        If int, sets the maximum number of thread to `limits` for each C-lib
+        selected by `subset`.
+
+        If dict(supported_libraries: max_threads), sets a custom maximum number
+        of thread for each C-lib.
+
+    subset : string or None, optional (default="all")
+        Subset of C-libs to limit. Used only if `limits` is an int
+
+        "all" : limit all supported C-libs.
+
+        "blas" : limit only BLAS supported C-libs.
+
+        "openmp" : limit only OpenMP supported C-libs. It can affect the number
+                   of threads used by the BLAS C-libs if they rely on OpenMP.
+    """
+    old_limits = get_thread_limits()
+    set_thread_limits(limits=limits, subset=subset)
+
+    try:
+        yield
```

Comment:
it is not clear to why `yield` and not `return`

Reply:
It's a context manager, not a simple function